jobs/etcd/templates/bin/authentication-setup.erb

```diff
@@ -0,0 +1,188 @@
+    if [ -z "$($etcdctl --user root:$ROOT_PASSWORD user get $username)" ]; then
+      $etcdctl --user root:$ROOT_PASSWORD user add $username:$password --interactive=false
+    fi
+  else
+    if [ -z "$($etcdctl --username root:$ROOT_PASSWORD user get $username)" ]; then
+      echo "$password" | $etcdctl --username root:$ROOT_PASSWORD user add $username
+    fi
+  fi
+}
+
+etcdctl_add_role() {
+  role=$1
+
+  if [ "$ETCDCTL_API" -eq 3 ]; then
+    if [ -z "$($etcdctl --user root:$ROOT_PASSWORD role get $role)" ]; then
+      $etcdctl --user root:$ROOT_PASSWORD role add $role
+    fi
+  else
+    if [ -z "$($etcdctl --username root:$ROOT_PASSWORD role get $role)" ]; then
+      $etcdctl --username root:$ROOT_PASSWORD role add $role
+    fi
+  fi
+}
+
+etcdctl_grant_read_permission() {
+  role=$1
+  path=$2
+  read_regex=".*KV Read:.*\s+"$path".*KV Write:.*"
+
+  if [ "$ETCDCTL_API" -eq 3 ]; then
+    $etcdctl --user root:$ROOT_PASSWORD role grant-permission $role read "$path"
+  else
+    if ! [[ $($etcdctl --username root:$ROOT_PASSWORD role get $role) =~ $read_regex ]]; then
+      $etcdctl --username root:$ROOT_PASSWORD role grant $role --read --path "$path"
+    fi
+  fi
+}
+
+etcdctl_grant_write_permission() {
+  role=$1
+  path=$2
+  write_regex=".*KV Write:.*\s+"$path"\s+.*|.*KV Write:.*\s+"$path
+
+  if [ "$ETCDCTL_API" -eq 3 ]; then
+    read_regex=".*KV Read:.*\s+${path}KV Write:.*|.*KV Read:.*\s+$path\s+.*KV Write:.*"
+    if [[ $($etcdctl --user root:$ROOT_PASSWORD role get $role) =~ $read_regex ]]; then
+      $etcdctl --user root:$ROOT_PASSWORD role grant-permission $role readwrite "$path"
+    else
+      $etcdctl --user root:$ROOT_PASSWORD role grant-permission $role write "$path"
+    fi
+  else
+    if ! [[ $($etcdctl --username root:$ROOT_PASSWORD role get $role) =~ $write_regex ]]; then
+      $etcdctl --username root:$ROOT_PASSWORD role grant $role --write --path "$path"
+    fi
+  fi
+}
+
+etcdctl_bind_role_to_user() {
+  username=$1
+  role=$2
+  role_regex=".*Roles:.*\s+"$role"\s+.*|.*Roles:.*\s+"$role
+
+  if [ "$ETCDCTL_API" -eq 3 ]; then
+    $etcdctl --user root:$ROOT_PASSWORD user grant-role $username $role
+  else
+    if ! [[ "$($etcdctl --username root:$ROOT_PASSWORD user get $username)" =~ $role_regex ]]; then
+      $etcdctl --username root:$ROOT_PASSWORD user grant $username --roles $role
+    fi
+  fi
+}
+
+show_user_role() {
+  username=$1
+  role=$2
+
+  if [ "$ETCDCTL_API" -eq 3 ]; then
+    $etcdctl --user root:$ROOT_PASSWORD user get $username
+    $etcdctl --user root:$ROOT_PASSWORD role get $role
+  else
+    $etcdctl --username root:$ROOT_PASSWORD user get $username
+    $etcdctl --username root:$ROOT_PASSWORD role get $role
+  fi
+}
+
+setup_users_auth() {
+  <% if_p('users') do |users| %>
+  versions=""
+  <% users.each do |user| %>
+    <% if user["versions"] != nil %>
+    <% user["versions"].each do |version| %>
+      <% if version == "v3" %>
+  export ETCDCTL_API=3
```

Comment:
Any chance we can do this without mucking with the exported ENV? It's akin to using a global variable and makes it a little difficult to track what's going on here.